### PR TITLE
[Relay] Fix operator fusion for multiple output

### PR DESCRIPTION
**Pull request by MarisaKirisame**
See #3867.

@tqchen @masahi @vinx13 

I cant get a test case as it depend on the gradient of a concat which is another WIP. I could add the test afterward though.

### CODE REVIEW

**Comment by vinx13** [comment]
Can you manually construct a Relay program for test?

**Comment by MarisaKirisame** [comment]
@vinx13 I can try. No grantee though as the failure case is generated and complex.

**Comment by MarisaKirisame** [comment]
@vinx13 the test case was found.

File: src/relay/ir/pretty_printer.cc [hunk 1]
```diff
@@ -304,14 +304,16 @@ class PrettyPrinter :
    * \return The corresponding name.
    */
   Doc AllocTypeVar(const TypeVar& var) {
+    if (memo_type_.count(var)) {
+      Doc val = memo_type_[var];
+      val << "-malformed-ir";
```
Comment (jroesch) [inline src/relay/ir/pretty_printer.cc]: Could we change this to print in a more informative way? for example maybe use a colored highlight to show which part is malformed?

Comment (MarisaKirisame) [reply src/relay/ir/pretty_printer.cc]: there is no color highlighting in the current doc.

File: src/relay/pass/fuse_ops.cc [hunk 2]
```diff
@@ -863,7 +865,7 @@ class FuseMutator : private ExprMutator {
     auto* ret_group = gmap_.at(tuple_get)->FindRoot();
     auto new_tuple = GetNewArguments({tuple_get->tuple}, ret_group)[0];
     auto new_node = TupleGetItemNode::make(new_tuple, tuple_get->index);
-    if (ret_group == gmap_.at(tuple_get)) {
+    if (ret_group->root_ref == tuple_get) {
```
Comment (masahi) [inline src/relay/pass/fuse_ops.cc]: why do we need this change? Aren't these conditions equivalent?

Comment (MarisaKirisame) [reply src/relay/pass/fuse_ops.cc]: this is only due to consistency to the same case on Call.

Comment (MarisaKirisame) [reply src/relay/pass/fuse_ops.cc]: (a small refactoring)

Comment (masahi) [reply src/relay/pass/fuse_ops.cc]: I see, then you can also update [the same condition in Tuple](https://github.com/dmlc/tvm/blob/master/src/relay/pass/fuse_ops.cc#L854) too.

Comment (MarisaKirisame) [reply src/relay/pass/fuse_ops.cc]: good catch.

Comment (MarisaKirisame) [reply src/relay/pass/fuse_ops.cc]: done.

File: src/relay/pass/fuse_ops.cc [hunk 3]
```diff
@@ -850,8 +852,8 @@ class FuseMutator : private ExprMutator {
   }
 
   Expr VisitExpr_(const TupleNode* tuple) {
-    auto* ret_group = gmap_.at(tuple)->FindRoot();
-    if (ret_group == gmap_.at(tuple)) {
+      auto* ret_group = gmap_.at(tuple)->FindRoot();
+      if (ret_group->root_ref == tuple) {
```
Comment (masahi) [inline src/relay/pass/fuse_ops.cc]: Extra spaces?

**Comment by masahi** [comment]
Here is my understanding of the problem, based on @MarisaKirisame's testcast:

- The Split node has multiple outputs. All outputs meet at the add node at the end, but since one of the output involves reference, they cannot be fused into a single op
- So the split node doesn't have a parent in the postdom tree.
- But before this PR, the split node gets one of its outputs node as its parent no matter what. 

Is this correct? @MarisaKirisame

**Comment by MarisaKirisame** [comment]
@masahi yes. the gradient of concatenate (which is split) triggered this problem.

**Comment by masahi** [comment]
Can you make a new function that return the LCA of all output nodes? In this case the new function should return null. I think this is clearer and less ad hoc.

**Comment by MarisaKirisame** [comment]
@masahi like this?

**Comment by masahi** [comment]
I mean, I want to make the following snippet easier to understand. I know your fix works, but it took me a while to understand what problem this PR is supposed to solve. Especially the logic around assigning the parent pointer temporarily and letting LCA fix up when the parent should be null seems tricky.

```
// find the LCAs of all outputs.
OpPatternKind pattern = kElemWise;
Node* parent = nullptr;
bool init = true;
for (auto link = gnode->outputs.head; link != nullptr; link= link->next) {
  size_t oindex = link->value.node->index;
  CHECK_LT(oindex, nodes.size());
  Node* onode = nodes[oindex];
  CHECK(onode != nullptr);
  if (init) {
    parent = onode;
    init = false;
  } else {
    parent = LeastCommonAncestor(parent, onode, &pattern);
  }
  pattern = CombinePattern(pattern, link->value.pattern);
}
```

It would be great if we could replace snippet above with something like (using a mix of non c++ syntax):
```
parent, pattern = LeastCommonAncestor(gnode->outputs) // LCA of all output nodes, not just two
tnode->parent = parent;
tnode->depth = parent ? parent->depth + 1 : 1;
tnode->pattern = pattern;
```

**Comment by masahi** [comment]
How about something like this? It passes your test but fails on other fusion tests :)

```
  static std::pair<DominatorTree::Node*, OpPatternKind> 
  LeastCommonAncestor(const LinkedList<IndexedForwardGraph::Edge>& outputs,
                      const DominatorTree& tree) {
    std::vector<DominatorTree::Node*> nodes;
    for (auto link = outputs.head; link != nullptr; link = link->next) {
      size_t oindex = link->value.node->index;
      CHECK_LT(oindex, tree.nodes.size());
      auto* onode = tree.nodes[oindex];
      CHECK(onode != nullptr);
      nodes.push_back(onode);
    }
    OpPatternKind pattern = kElemWise;
    if (nodes.empty()) return std::make_pair(nullptr, pattern);
    auto combine_func = [=, &pattern](DominatorTree::Node* n1, DominatorTree::Node* n2) {
      return LeastCommonAncestor(n1, n2, &pattern);
    };
    auto lca = std::accumulate(nodes.begin(), nodes.end(), nodes[0], combine_func);
    return std::make_pair(lca, pattern);
  }
```

**Comment by MarisaKirisame** [comment]
@masahi you have to treat the first node specially. You are not really doing a accumulate, but you are:
0: interleaving between all the children
1: if there is no children, provide a default value.

**Comment by masahi** [comment]
hmm I update the code below and now it passes all fusion test cases. Not sure if I am doing something wrong.

```
static std::pair<DominatorTree::Node*, OpPatternKind> 
  LeastCommonAncestor(const LinkedList<IndexedForwardGraph::Edge>& outputs,
                      const DominatorTree& tree) {
    std::vector<std::pair<DominatorTree::Node*, OpPatternKind>> nodes;
    for (auto link = outputs.head; link != nullptr; link = link->next) {
      size_t oindex = link->value.node->index;
      CHECK_LT(oindex, tree.nodes.size());
      auto* onode = tree.nodes[oindex];
      CHECK(onode != nullptr);
      nodes.push_back(std::make_pair(onode, link->value.pattern));
    }
    if (nodes.empty()) return std::make_pair(nullptr, kElemWise);
    auto combine_func = [=](std::pair<DominatorTree::Node*, OpPatternKind> accum,
                            std::pair<DominatorTree::Node*, OpPatternKind> next) {
      return std::make_pair(
          LeastCommonAncestor(std::get<0>(accum), std::get<0>(next), &std::get<1>(accum)),
          CombinePattern(std::get<1>(accum), std::get<1>(next)));
    };
    auto lca = std::accumulate(nodes.begin(), nodes.end(), nodes[0], combine_func);
    return lca;
  }
```

**Comment by MarisaKirisame** [comment]
@masahi I just wrote one that does basically the same thing. How is it?

**Comment by MarisaKirisame** [comment]
Also your std::get<1>(accum) and &std::get<1>(accum) will create very subtle errors that depend on the order of which execute first.

**Comment by masahi** [comment]
Yeah it looks better. Thanks :)

**Comment by jroesch** [comment]
@masahi can you approve explicitly? I can then merge then

**Comment by masahi** [comment]
@jroesch sorry I was waiting for the CI to finish :)

**Pull request merged**